src_c/font.c

```diff
@@ -388,121 +387,83 @@ font_render(PyObject *self, PyObject *args)
         /* Exception already set for us */
         return NULL;
     }
-    foreg.r = rgba[0];
-    foreg.g = rgba[1];
-    foreg.b = rgba[2];
-    foreg.a = SDL_ALPHA_OPAQUE;
-    if (bg_rgba_obj == Py_None) {
-        /* Explicit None is the same as not passing a color for us */
-        bg_rgba_obj = NULL;
-    }
-    if (bg_rgba_obj != NULL) {
+
+    SDL_Color foreg = {rgba[0], rgba[1], rgba[2], SDL_ALPHA_OPAQUE};
+    /* might be overriden right below, with an explicit background color */
+    SDL_Color backg = {0, 0, 0, SDL_ALPHA_OPAQUE};
+
+    if (bg_rgba_obj != Py_None) {
         if (!pg_RGBAFromFuzzyColorObj(bg_rgba_obj, rgba)) {
             /* Exception already set for us */
             return NULL;
         }
-        else {
-            backg.r = rgba[0];
-            backg.g = rgba[1];
-            backg.b = rgba[2];
-            backg.a = SDL_ALPHA_OPAQUE;
-        }
-    }
-    else {
-        backg.r = 0;
-        backg.g = 0;
-        backg.b = 0;
-        backg.a = SDL_ALPHA_OPAQUE;
+        backg = (SDL_Color){rgba[0], rgba[1], rgba[2], SDL_ALPHA_OPAQUE};
     }
 
-    just_return = PyObject_Not(text);
-    if (just_return) {
-        int height = TTF_FontHeight(font);
-
-        if (just_return == -1 || !(PyUnicode_Check(text) ||
-                                   PyBytes_Check(text) || text == Py_None)) {
-            PyErr_Clear();
-            return RAISE_TEXT_TYPE_ERROR();
-        }
-        surf = SDL_CreateRGBSurface(SDL_SWSURFACE, 0, height, 32, 0xff << 16,
-                                    0xff << 8, 0xff, 0);
-        if (surf == NULL) {
-            return RAISE(pgExc_SDLError, SDL_GetError());
-        }
-        if (bg_rgba_obj != NULL) {
-            Uint32 c = SDL_MapRGB(surf->format, backg.r, backg.g, backg.b);
-            SDL_FillRect(surf, NULL, c);
-        }
-        else {
-            SDL_SetColorKey(surf, SDL_SRCCOLORKEY, 0);
-        }
+    if (!PyUnicode_Check(text) && !PyBytes_Check(text) && text != Py_None) {
+        return RAISE_TEXT_TYPE_ERROR();
     }
-    else if (PyUnicode_Check(text)) {
-        PyObject *bytes = PyUnicode_AsEncodedString(text, "utf-8", "replace");
-        const char *astring = NULL;
 
-        if (!bytes) {
+    if (PyUnicode_Check(text)) {
+        Py_ssize_t _size = -1;
+        astring = PyUnicode_AsUTF8AndSize(text, &_size);
+        if (astring == NULL) { /* exception already set */
             return NULL;
         }
-        astring = PyBytes_AsString(bytes);
-        if (strlen(astring) != (size_t)PyBytes_GET_SIZE(bytes)) {
-            Py_DECREF(bytes);
+        if (strlen(astring) != (size_t)_size) {
             return RAISE(PyExc_ValueError,
                          "A null character was found in the text");
         }
+    }
+
+    if (PyBytes_Check(text)) {
```

Comment:
```suggestion
    else if (PyBytes_Check(text)) {
```
We don't need to bother with a bytes check if `text` was already unicode